### DISCUSSION TITLE
Ver 4.1.2 - Phalcon is trying to load subdirectory as Controller

### DISCUSSION THREAD

**Discussion by RickyJaff** [phalcon/cphalcon#15495] [Q&A] [answered]
Hi, 

Installed phalcon 4.1.2 in windows using Uniform Server on PHP 7.4 . I am not able to get it work. 
I have installed in /www/phalcon4

For testing with "hello world" code ,

On loading "http://localhost/phalcon4" , I get following error: 
`Exception: Phalcon4Controller handler class cannot be load`

If I try same Hello world code on Phalcon 3.4 (installed on PHP 7.1) then it works. 

I see similar problem faced by someone here : https://forum.phalcon.io/discussion/20211/project-in-a-subdirectory  and  there was no solution and he was using some "WORKAROUND". 

In my case, since I have some experience with Phalcon in past (worked on 3.4 in on mid-sized project), all I can see is that Phalcon is thinking "subdirectory" from URL as controller and trying to find controller for it and hence error. 

Here is my www/phalcon4/.htaccess

```
Options +FollowSymLinks -MultiViews -Indexes

<IfModule mod_rewrite.c>
	RewriteEngine on
	RewriteRule  ^$ public/    [L]
	RewriteRule  (.*) public/$1 [L]
</IfModule>
```

Here is .htaccess for  www/phalcon4/public

```
AddDefaultCharset UTF-8

<IfModule mod_rewrite.c>
    RewriteEngine On
    RewriteCond %{REQUEST_FILENAME} !-d
    RewriteCond %{REQUEST_FILENAME} !-f
    RewriteRule ^(.*)$ index.php?_url=/$1 [QSA,L]
</IfModule>
```
In public/index.php  baseUri is set as following:
```

// Setup a base URI
$di->set(
    'url',
    function () {
        $url = new UrlProvider();
        $url->setBaseUri('/phalcon4/');
        return $url;
    }
);
```
In past, I have worked on dev setup in subfolder many times, but that was with phalcon 3.x , tried 4.1.2 yesterday first time and still stuck in this problem. 

In short, my question is : 

**How to make Phalcon 4 to work in subdirectory ?** 

Just in case, phalcon is working on my PHP :

![phalcon-7 4 16](https://user-images.githubusercontent.com/1543029/118692670-ecbd5b00-b827-11eb-93fb-9c0d7f221c8d.jpg)

**Comment by BeMySlaveDarlin** [comment, marked as answer]
First of all, thank you for report.
There are few ways to fix this, but as is see, you are using some custom server with built-in apache webserver.
1. Using apache you can trim `phalcon4` part of request url in line i marked next:
```
<IfModule mod_rewrite.c>
    RewriteEngine On
    RewriteCond %{REQUEST_FILENAME} !-d
    RewriteCond %{REQUEST_FILENAME} !-f
    RewriteRule ^(.*)$ index.php?_url=/$1 [QSA,L] # here
</IfModule>
```
2. Or trim `/phalcon4` part of request url in `index.php`:
```
    //default
    echo $application->handle($_SERVER['REQUEST_URI'])->getContent();
```
```
    //fixed
    $requestUri = str_replace('phalcon4', '', $_SERVER['REQUEST_URI']);
    echo $application->handle($requestUri)->getContent();
```
Editing `url` service wont help, because that affects only generated urls you will render later.

**Reply by RickyJaff** [reply]
Thanks for the reply, following hack works as now Phalcon is not seeing that subdirectory in request and hence not considering it. 

```
    //fixed
    $requestUri = str_replace('phalcon4', '', $_SERVER['REQUEST_URI']);
    echo $application->handle($requestUri)->getContent();
```
However, this was never the case with 3.x version, should be count this as bug ?

**Reply by BeMySlaveDarlin** [reply]
The way 3.x routes request - is obsolete and wrong.
New logic is not a bug.
